Fix inconsistent behaviour when a charset is defined in Content-Type header value with an invalid and valid json #1202

scoulomb opened the pull request:
# Fixes 
Issue https://github.com/zalando/connexion/issues/1202

scoulomb commented:
I tested it locally here:
https://github.com/scoulomb/zalando_connexion_sample  (+buildLocalLib branch)

I did not see specific unit test to update, what would you advise?

Jiehong commented on connexion/utils.py:
```diff
@@ -131,7 +131,9 @@ def is_json_mimetype(mimetype):
-    return maintype == 'application' and (subtype == 'json' or subtype.endswith('+json'))
+    return maintype == 'application' and (subtype == 'json'
+                                          or subtype == 'json; charset=utf-8'
+                                          or subtype.endswith('+json'))
```
Would using `subtype.containes('json')` be enough to better cover this?

scoulomb replied:
If doing this we would accept invalid content-type? like `Content-type: application/json; charset=utf-83w9fmsefew0`, 

Also from https://stackoverflow.com/questions/9254891/what-does-content-type-application-json-charset-utf-8-really-mean and http://www.ietf.org/rfc/rfc4627.txt. 
We can see `Content-type: application/json; charset=utf-8 ` <=> `Content-type: application/json`, otherwise we would have ti support more encoding.

Jiehong replied:
For the RFC4627, you didn't even highlight a specific passage, but I still checked and saw 2 things:

     JSON text SHALL be encoded in Unicode.  The default encoding is
     UTF-8.

and 


      JSON may be represented using UTF-8, UTF-16, or UTF-32.  When JSON
      is written in UTF-8, JSON is 8bit compatible.  When JSON is
      written in UTF-16 or UTF-32, the binary content-transfer-encoding
      must be used.

So it is indeed possible to not only send UTF-8.

But in our case, I agree it's fine that utf-8 is enough, as it's the default for now.

How would the code deal with `application/json;charset=utf-8`?

scoulomb replied:
Based on our discussion, what do you think of this implementation?
```
# Reference is https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types
# Where we can see MIME type can have optional parameter
# We can have a space or not in charset
def is_json_mimetype(mimetype):
    """
    :type mimetype: str
    :rtype: bool
    """
    simple_mime_type, optional_parameter = mimetype.split(";") if \
        ";" in mimetype else \
        (mimetype, None)
    maintype, subtype = simple_mime_type.split('/')  # type: str, str

    return maintype == 'application' and \
           (
                   subtype == 'json' or
                   subtype.endswith('+json')
           ) and \
           (
                   optional_parameter is None or
                   optional_parameter.strip() == "charset=utf-8"
           )


```

where

```
import unittest

from mimetype import is_json_mimetype


class TestUtils(unittest.TestCase):

    def test_is_json_mimetype_nominal(self):

        print(is_json_mimetype("application/json"))
        self.assertTrue(is_json_mimetype("application/json"))
        self.assertTrue(is_json_mimetype("application/json; charset=utf-8"))
        self.assertTrue(is_json_mimetype("application/json;charset=utf-8"))
        self.assertTrue(is_json_mimetype("application/anything+json;charset=utf-8"))
        self.assertTrue(is_json_mimetype("application/anything+json"))

        self.assertFalse(is_json_mimetype("application/xml"))
        self.assertFalse(is_json_mimetype("application/json;charset=utf-dfdfd"))

````



`

Jiehong replied:
I think this is better. (Just remove the print in the test, and maybe simplify the return value of the function, as it's a 9-lines return value).

Go ahead!

scoulomb replied:
Thanks for your suggestion I have updated accordingly.
I made some test with a sample server to check behavior here:
https://github.com/scoulomb/connexion/commit/fc315835c1de37707835f9d1f8043588e255d35e

scoulomb commented:
Fix loss of coverage here: https://github.com/zalando/connexion/blob/c99dda68f4cad9e36815267eb68898220499ee5c/connexion/decorators/validation.py#L142

scoulomb commented:
I added unit test in  test_json_validation.py 
However loss of coverage was due to the fact that in the former version of the code we were doing tuple unpacking after the split which could lead to a ValueError exception

>>> mimetype.split('/')
['application-json']
>>> a,b = mimetype.split('/')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: not enough values to unpack (expected 2, got 1)

IMO it is a workaround to test a side effect of the function outsside of it:
As we now test separator presence it is useless to catch this exception so I will remove the try/except block

scoulomb commented:
Hello @hjacobs , I fixed https://github.com/zalando/connexion/issues/1202 in this PR. Would you mind please have a look? Thanks a lot.

hjacobs commented on connexion/utils.py:
```diff
@@ -131,8 +131,21 @@ def is_json_mimetype(mimetype):
     """
-    maintype, subtype = mimetype.split('/')  # type: str, str
-    return maintype == 'application' and (subtype == 'json' or subtype.endswith('+json'))
+    simple_mime_type, optional_parameter = mimetype.split(";") if \
```
you can simplify this code by using `mimetype.partition(";")` instead

hjacobs commented on connexion/utils.py:
```diff
@@ -131,8 +131,21 @@ def is_json_mimetype(mimetype):
+        (mimetype, None)
+    if "/" not in simple_mime_type:
+        return False
+    maintype, subtype = simple_mime_type.split('/')  # type: str, str
```
same here: `partition(..)` will always return 3 values, so we can get rid of the `if`

scoulomb replied:
@hjacobs Very elegant. Thanks. I updated.

scoulomb commented:
@hjacobs do you think this PR is mergeable, or should I rework on it?

RobbeSneyders commented:
This seems like a good change @scoulomb. If you're still interested in getting it merged, please rebase on and target this PR to the `v2` branch.
